File: bench/benches/strscan.js
// fixed string built once; the timed work is a linear rolling-hash scan.
const { bench } = require("../lib/bench");
const HMOD = 1000000007;
const data = Array.from({ length: 20000 }, (_, i) => String.fromCharCode(32 + (7 * i) % 95)).join("");
bench("strscan", () => {
  let h = 0;
  for (let i = 0; i < data.length; i++) h = (h * 31 + data.charCodeAt(i)) % HMOD;
  return h;
});

File: bench/benches/strscan.rs
include!("../lib/bench.rs");

const HMOD: i64 = 1000000007;

// fixed string built once; the timed work is a linear rolling-hash scan.
// chars are ASCII (32 + (7*i)%95) in 32..126, single-byte.
fn main() {
    let data: Vec<u8> = (0..20000i64).map(|i| (32 + (7 * i) % 95) as u8).collect();
    bench("strscan", || {
        let mut h: i64 = 0;
        for &ch in &data {
            h = (h * 31 + ch as i64) % HMOD;
        }
        h
    });
}
